src/masternodes/rpc_accounts.cpp

```diff
@@ -1074,108 +1074,82 @@ UniValue listaccounthistory(const JSONRPCRequest& request) {
+    std::set<CScript> rewardAccounts;
+    std::queue<CRewardHistory> rewardsHistory;
 
     auto shouldContinueToNextAccountHistory = [&](AccountHistoryKey const & key, CLazySerialize<AccountHistoryValue> valueLazy) -> bool {
-        if (!isMatchOwner(key.owner)) {
-            return false;
-        }
 
-        std::unique_ptr<CScopeAccountReverter> reverter;
-        if (!noRewards) {
-            reverter = MakeUnique<CScopeAccountReverter>(view, key.owner, valueLazy.get().diff);
+        if (startBlock > key.blockHeight) {
+            return false;
         }
 
-        bool accountRecord = true;
-        auto workingHeight = key.blockHeight;
-
-        if (shouldSkipBlock(key.blockHeight)) {
-            // show rewards in interval [startBlock, lastHeight)
-            if (!noRewards && startBlock > workingHeight) {
-                accountRecord = false;
-                workingHeight = startBlock;
-            } else {
-                return true;
-            }
+        if (!isMatchOwner(key.owner)) {
+            return true;
         }
 
         if (isMine && !(IsMineCached(*pwallet, key.owner) & filter)) {
             return true;
         }
 
-        const auto & value = valueLazy.get();
-
-        if (CustomTxType::None != txType && value.category != uint8_t(txType)) {
+        bool accountRecord = maxBlockHeight >= key.blockHeight;
+        if (!accountRecord && noRewards) {
             return true;
         }
 
-        if (isMine) {
-            // starts new account owned by the wallet
-            if (lastOwner != key.owner) {
-                count = limit;
-            } else if (count == 0) {
-                return true;
-            }
+        const auto & value = valueLazy.get();
+
+        if (!noRewards) {
+            auto& owner = *(rewardAccounts.insert(key.owner).first);
+            rewardsHistory.push({key.blockHeight, owner, value.diff});
         }
 
-        // starting new account
-        if (account.empty() && lastOwner != key.owner) {
-            view.Discard();
-            lastOwner = key.owner;
-            lastHeight = maxBlockHeight;
+        if (CustomTxType::None != txType && value.category != uint8_t(txType)) {
+            return true;
         }
 
         if (accountRecord && (tokenFilter.empty() || hasToken(value.diff))) {
-            auto& array = ret.emplace(workingHeight, UniValue::VARR).first->second;
+            auto& array = ret.emplace(key.blockHeight, UniValue::VARR).first->second;
             array.push_back(accounthistoryToJSON(key, value));
             if (shouldSearchInWallet) {
                 txs.insert(value.txid);
             }
             --count;
         }
 
-        if (!noRewards && count && lastHeight > workingHeight) {
-            onPoolRewards(view, key.owner, workingHeight, lastHeight,
+        return count != 0;
+    };
+
+    paccountHistoryDB->ForEachAccountHistory(shouldContinueToNextAccountHistory, {~0u, {}, ~0u});
```

Comment:
This is equivalent to a full scan as I can see. Do not make use of the index properly. earlier it was like
```
AccountHistoryKey startKey{account, maxBlockHeight, std::numeric_limits<uint32_t>::max()};
paccountHistoryDB->ForEachAccountHistory(shouldContinueToNextAccountHistory, startKey);
```
so the iteration starts from a given `startKey`. now it starts from one end, and iterates, filters within the lambda.
I think we could pass the starting key?🤔

Also, this is important for `Whale` because it depends on the old index `<owner, height,  txn>` to get max performance by passing the `owner`. Now the index has been changed to `<height, owner, txn>` it has to be taken into account @fuxingloh @canonbrother @ivan-zynesis

Reply:
Hmmm, we plan to expose this endpoint for `Whale` for Light Wallet and Scan users.

However, if this greatly benefits the Full Node users, I can scrap that endpoint for now. 🤔